### ISSUE TITLE
name resolver fails badly

### ISSUE THREAD

**Issue by kglotfelty** [cxcsds/ciao-contrib#703]
I'm not sure why we are seeing this now (maybe something with the fall-through to the secondary name server).

This is happening with 4.14 and 4.15 on mac (Monterey), ciao-install editions (so python 3.8 and 3.9), x86.

```bash
% find_chandra_obsid "FOOBAR" 
# find_chandra_obsid (27 January 2022): ERROR 'str' object has no attribute 'errno'


% find_chandra_obsid "FOOBAR" --debug verbose=5
Running: find_chandra_obsid
  version: 27 January 2022
Querying name resolvers for name=FOOBAR
CADC name query: https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/cadc-target-resolver/find?format=ascii&service=all&cached=true&target=FOOBAR
Error opening URL: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)>
error.reason = [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)
error.reason.errno = 1
Skipping to an unverified SSL context
Unable to query with no context!
Error opening URL: HTTP Error 425: 
error.reason = 
# find_chandra_obsid (27 January 2022): ERROR 'str' object has no attribute 'errno'


##### Traceback:

Traceback (most recent call last):
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 1354, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 950, in send
    self.connect()
  File "/Applications/ciao-4.14/ots/lib/python3.8/http/client.py", line 1424, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "/Applications/ciao-4.14/ots/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/Applications/ciao-4.14/ots/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/Applications/ciao-4.14/ots/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/coords/resolver.py", line 108, in query_url
    return urlopen(url)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 1397, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 1357, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/coords/resolver.py", line 138, in query_url
    return urlopen(url, context=context)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 563, in error
    result = self._call_chain(*args)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 755, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Applications/ciao-4.14/ots/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 425: 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/ciao_contrib/logger_wrapper.py", line 560, in new_fn
    return fn(*args, **kw)
  File "/Applications/ciao-4.14/contrib/bin/find_chandra_obsid", line 579, in fco
    run_fco(opts)
  File "/Applications/ciao-4.14/contrib/bin/find_chandra_obsid", line 538, in run_fco
    (ra, dec, csys) = resolver.identify_name(name)
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/coords/resolver.py", line 335, in identify_name
    return identify_name_cadc(name)
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/coords/resolver.py", line 185, in identify_name_cadc
    rsp = query_url(name, url, "CADC")
  File "/Applications/ciao-4.14/contrib/lib/python3.8/site-packages/coords/resolver.py", line 143, in query_url
    v5(f"error.reason.errno = {ue2.reason.errno}")
AttributeError: 'str' object has no attribute 'errno'
```

On Linux, it produces a useful error message:

```bash
% find_chandra_obsid "foobar" 
# find_chandra_obsid (27 January 2022): ERROR No position found matching the name 'foobar'.
```

**Comment by DougBurke** [comment]
Fooey...

**Comment by kglotfelty** [comment]
Well, it behaves nicely on my work CentOS7 box, but on my home Ubuntu 22.04 I get the same error as on the mac

```bash
find_chandra_obsid "Collinder197"
# find_chandra_obsid (27 January 2022): ERROR 'str' object has no attribute 'errno'
```

**Comment by kglotfelty** [comment]
no clue.  I tried the work around here: https://cxc.cfa.harvard.edu/ciao/bugs/python.html, but it didn't work.

On CentOS that works here's the debug+verbose=5 output

```bash
% find_chandra_obsid foobar --debug verbose=5
Running: find_chandra_obsid
  version: 27 January 2022
Querying name resolvers for name=foobar
CADC name query: https://www.cadc-ccda.hia-iha.nrc-cnrc.gc.ca/cadc-target-resolver/find?format=ascii&service=all&cached=true&target=foobar
Error from CADC name resolver - status = 425
Sesame name query: https://cdsweb.u-strasbg.fr/cgi-bin/nph-sesame/-ox/~SNV?foobar
Response from query:
<?xml version="1.0" encoding="UTF-8"?>
<Sesame xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 xsi:noNamespaceSchemaLocation="http://vizier.u-strasbg.fr/xml/sesame_4x.xsd">
<Target option="~SNV">
  <name>foobar</name>
  <!-- Q15829080 #1 -->
  <INFO> *** NNNothing found *** </INFO>
</Target>
</Sesame>
<!--- ====Done (2022-Dec-06,20:57:45z)==== -->

Parsing response as XML
# find_chandra_obsid (27 January 2022): ERROR No position found matching the name 'foobar'.


##### Traceback:

Traceback (most recent call last):
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 108, in query_url
    return urlopen(url)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 555, in error
    result = self._call_chain(*args)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 747, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/ots/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 425: 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 336, in identify_name
    return identify_name_cadc(name)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 185, in identify_name_cadc
    rsp = query_url(name, url, "CADC")
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 116, in query_url
    raise ValueError(f"No position found matching the name '{name}'.")
ValueError: No position found matching the name 'foobar'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/ciao_contrib/logger_wrapper.py", line 560, in new_fn
    return fn(*args, **kw)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/bin/find_chandra_obsid", line 579, in fco
    run_fco(opts)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/bin/find_chandra_obsid", line 538, in run_fco
    (ra, dec, csys) = resolver.identify_name(name)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 341, in identify_name
    return identify_name_sesame(name)
  File "/export/CIAOT/ciaot_install/20221122/ciao-4.15/contrib/lib/python3.9/site-packages/coords/resolver.py", line 274, in identify_name_sesame
    raise ValueError("No position found matching the name '{0}'.".format(name))
ValueError: No position found matching the name 'foobar'.
```

**Comment by kglotfelty** [comment]
Looking at https://docs.python.org/3/library/urllib.error.html?highlight=urlerror#urllib.error.URLError docs and the tracebacks, it looks like a `HTTPError` is being caught (derived from `URLError`).  Instead of  `ue2.reason.error` it looks like it should be using `ue2.code` ?

**Comment by DougBurke** [comment]
I'm assuming this is still an issue

**Comment by kglotfelty** [comment]
I didn’t see it this morning when I was reviewing things so I assume whatever was happening with the primary server was fixed on their end. I”m not sure if the issue is still there when it falls over to the backup server.

Kenny

**Comment by DougBurke** [comment]
Assuming fixed by #719
